pkg/routing/inmemory/inmemory.go

```diff
@@ -77,24 +79,62 @@ func (r *NodeInfoStore) Add(ctx context.Context, nodeInfo models.NodeInfo) error
 	return nil
 }
 
-func (r *NodeInfoStore) Get(ctx context.Context, peerID peer.ID) (models.NodeInfo, error) {
+func (r *NodeInfoStore) Get(ctx context.Context, nodeID string) (models.NodeInfo, error) {
 	r.mu.RLock()
 	defer r.mu.RUnlock()
-	infoWrapper, ok := r.nodeInfoMap[peerID]
+	infoWrapper, ok := r.nodeInfoMap[nodeID]
 	if !ok {
-		return models.NodeInfo{}, routing.NewErrNodeNotFound(peerID)
+		return models.NodeInfo{}, routing.NewErrNodeNotFound(nodeID)
 	}
 	if time.Now().After(infoWrapper.evictAt) {
 		go r.evict(ctx, infoWrapper)
-		return models.NodeInfo{}, routing.NewErrNodeNotFound(peerID)
+		return models.NodeInfo{}, routing.NewErrNodeNotFound(nodeID)
 	}
 	return infoWrapper.NodeInfo, nil
 }
 
+func (r *NodeInfoStore) GetByPrefix(ctx context.Context, prefix string) (models.NodeInfo, error) {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	nodeInfo, err := r.Get(ctx, prefix)
+
+	// we found a node with the exact ID
+	if err == nil {
+		return nodeInfo, nil
+	}
+	// return the error if it's not a node not found error
+	var errNotFound routing.ErrNodeNotFound
+	if !errors.As(err, &errNotFound) {
+		return models.NodeInfo{}, err
+	}
+
+	// look for a node with the prefix. if there are multiple nodes with the same prefix, return ErrMultipleNodesFound error
+	var nodeIDsWithPrefix []string
+	for nodeID, infoWrapper := range r.nodeInfoMap {
+		if time.Now().After(infoWrapper.evictAt) {
+			go r.evict(ctx, infoWrapper)
```

Comment:
This takes variadic params so we could collect the items be evicted and evict them in a single goroutine.

Reply:
Fixed in https://github.com/bacalhau-project/bacalhau/pull/2986/commits/590a38684f703bb33e27ecfeb9e1e739ee198cbe

as for the 500 error, almost all failed calls return 500 in the new APIs. I have a backlog item (still in Jira) to address this and improve error reporting